### ISSUE TITLE
`sudo apt install <tab><tab>` freeze my terminal

### ISSUE THREAD

**Issue by madfatcat88** [ohmyzsh/ohmyzsh#10810]
### Describe the bug

when I try to run `sudo apt install <tab><tab>`, my ternimal just freeze

### Steps to reproduce

1. install zsh
2. run `sudo apt install <tab><tab>` -> auto complete works just find
3. install ohmyzh
4. run `sudo apt install <tab><tab>` -> terminal freeze and it show up as `sudo apt install 0ad` (0ad was automatically added to the end), can not do anything for a while

### Expected behavior

It should work just as before installing ohmyzsh

### OS / Linux distribution

Linux mint 20.3, Ubuntu server 20.4

### Zsh version

5.8

### Oh My Zsh version

master (4dce175)

### Terminal emulator

tilix

**Comment by mcornella** [comment]
> * install zsh
> * run `sudo apt install <tab>` -> auto complete works just find

What suggestions do you get here? I presume you should only get files instead of packages no?

Oh My Zsh does nothing for apt install completion, it only enables zsh completion so what you're getting is the available packages for install. `0ad` just happens to be the first one alphabetically, but you should get the list for the rest of available packages as well.

**Comment by madfatcat88** [comment]
> What suggestions do you get here? I presume you should only get files instead of packages no?

I get package suggestions.


> `0ad` just happens to be the first one alphabetically, but you should get the list for the rest of available packages as well

yes, but it should not be appended to the command line that I input. And yes, I still get the list, but it freezes my terminal for a few minutes before I can get the list and this never happens with bash or zsh minus OMZ.

**Comment by madfatcat88** [comment]
I think I get your idea about the `0ad` and I agree that autocomplete fills `0ad` to the command line, no problem here.
The only thing that troubles me is the hanging terminal, it blocks all kind of interaction (CTRL C, CTRL D...)

**Comment by mcornella** [comment]
Can you capture a trace of the completion process for zsh-only and with OMZ? You can do that by typing the command (`sudo apt install `), then pressing <kbd>CTRL+X</kbd> <kbd>?</kbd>.

This should give you something like
![image](https://user-images.githubusercontent.com/1441704/160592949-e982024a-ed37-49cf-b123-a7a825ef2760.png)

You have to then upload the `/tmp/xxxx` files.

**Comment by madfatcat88** [comment]
I got nothing when pressing `CTRL + X`.
But I noticed that my steps to reproduce was missing a little detail. it should be `sudo apt install <tab><tab>`, 2 tabs continuously instead of 1.
Sorry, it was my bad.

**Comment by mcornella** [comment]
You should press `CTRL-X` and then `?`. Sorry if that wasn't clear.

**Comment by madfatcat88** [comment]
Here they are
[zsh_with_omz.log](https://github.com/ohmyzsh/ohmyzsh/files/8394218/zsh_with_omz.log)
[zsh_no_omz.log](https://github.com/ohmyzsh/ohmyzsh/files/8394220/zsh_no_omz.log)

**Comment by mcornella** [comment]
So it looks like in OMZ the completion correctly uses the cache, so that should be faster, not slower. You should also not have to type TAB twice, maybe that's why you get `0ad` selected (try pressing it only once and wait for the results to come in).

In the log file you posted from with OMZ, it looks like the cache was already retrieved, so I can't see what happens during that execution. In any case, all that does is source a cache file, so it shouldn't cause any delays.

I'd love it if you can also upload the cache file, you should be able to find at `$ZSH_CACHE_DIR/DEBS_avail`.

Then let's try to remove the file so it is recreated on the next completion. You should restart zsh afterwards as well, so run this:

```zsh
rm $ZSH_CACHE_DIR/DEBS_avail
omz reload
```
Here `omz reload` is to restart the zsh session, functionally equivalent to `exec zsh`.

After that, try to complete `sudo apt install` again, and let me know if the speed issues are still there.

**Comment by madfatcat88** [comment]
here's the cache file. And it still freezes after I removed the file and reload the shell.
[DEBS_avail.log](https://github.com/ohmyzsh/ohmyzsh/files/8398517/DEBS_avail.log)

**Comment by mcornella** [comment]
That looks fine. 

Let's see if the completion stops at any point in particular. First remove the file and reload the session again. Then run `functions -t ${_comps[apt]}` to trace the completion function for `apt`, and then try to complete the command again.

You should be able to see what lines printed to the console as the function runs. Let me know if it stops at any point in particular.

Oh BTW, is `$ZSH_CACHE_DIR` in a network path, or some other filesystem that is slow? That could explain the slowness of accessing the cache file.

**Comment by madfatcat88** [comment]
Here's what I can get
[sudo_apt_zsh.log](https://github.com/ohmyzsh/ohmyzsh/files/8399400/sudo_apt_zsh.log)
I have 3 PC. and all of them run on SSDs and all have the same issue.
I understand that 1 `tab` is enough, but it's my habit to hit it twice. I try to change it but it happens sometimes, still.
 
Also, I notice something else. If I type `sudo apt install <tab>` then it's gonna ask me `zsh: do you wish to see all 71394 possibilities (71394 lines)? ` and if I hit `tab` again, it's gonna print all `71394 lines` on the screen and it takes about 30s for that. So I think then I type `sudo apt install <tab><tab>` then it would do the same thing, the only thing different is that it does not show the whole 71394 lines on the screen but load it into memory and once the loading is done, it will display the list as a selection list below the prompt so that I can select the item with my arrow key.
Bottom line,  I think there's something fishy that happens when OMZ load the long list and displayit it as a selection list right below the prompt.

**Comment by madfatcat88** [comment]
Any news?

**Comment by madfatcat88** [comment]
Still have this problem with the latest OMZ, pretty annoying.

**Comment by aexkcz** [comment]
I got the same problem. still no solution?

**Comment by madfatcat88** [comment]
seem so

**Comment by mcornella** [comment]
I have been researching this further, and this is expected behavior when pressing TAB twice and having no paginated completion.

You can sort of work around this by enabling paginated completion. You can use the settings below and put them on your zshrc file:
```zsh
zstyle ':completion:*' list-prompt   ''
zstyle ':completion:*' select-prompt ''
```

This will show all completions in a menu when there are more entries than fit in the page. You can scroll on it by pressing TAB.

Either way, the best approach is to be aware that there will be a lot of entries, and to never press TAB twice to complete a command (you can press TAB again to select the first entry, but don't do it as a default and you don't need to do it either to get completion). On a single TAB press, you also get a reminder of the number of entries:

```console
$ apt install 
zsh: do you wish to see all 57056 possibilities (28528 lines)? 
```
